### PR TITLE
fix(transformer): remove span of define value

### LINKED ISSUE #7797: `define` generates incorrect sourcemap
Reproduction: https://stackblitz.com/edit/github-xv4yss-vvnxe2dc?file=build.mjs
The reproduction uses `rolldown@0.15.0-snapshot-3cea4f5-20241211003613` which uses oxc@0.40.1.
image

### PR DESCRIPTION
**Pull request by hi-ogawa**
- Closes https://github.com/oxc-project/oxc/issues/7797

Removing span seems to work, but I'm not sure traversing all the time is okay.

### CODE REVIEW

**Comment by overlookmotel** [comment]
I don't think this will generalize well. As I understand it, the replacement expression can be anything, and this will not remove the spans from e.g. `(() => { return 123; })()`.

To solve this fully, we either need:

1. Option to parser to produce empty spans for all nodes.
2. `VisitMut` implementation which removes span from *all* nodes. This could be generated using `ast_tools` codegen.

The 1st option would be faster for this particular use case, but we'd need to make sure it doesn't produce a noticeable slowdown for parsing in "standard mode". We could make it a generic param `ParserImpl<const EMPTY_SPANS: bool>` but that'd produce a large binary size increase.

So probably 2nd option is better.

**Comment by hi-ogawa** [comment]
> As I understand it, the replacement expression can be anything, and this will not remove all the spans from e.g. `(() => { return 123; })()`.

I was wondering about that too. Esbuild doesn't allow arbitrary expression (probably it supports only json value, identifier, member expression https://esbuild.github.io/api/#define), but oxc can do different.

For the usage I'm aware of, I thought this could be a reasonable fix to start with, but if better fix is desired, I would be happy to try `ast_tools` approach as I'm interested in it (and I'm scared to touch parser code :sweat_smile:). Please let me know!

**Comment by overlookmotel** [comment]
#7816 adds a `visit_span` method to `Visit` and `VisitMut`. This would make implementing `RemoveSpans` trivial:

```rs
impl<'a> VisitMut<'a> for RemoveSpans {
  fn visit_span(&mut self, span: &mut Span) {
    *span = SPAN;
  }
}
```

Only problem is that #7816 may cause a perf regression. I *hope* it won't as `visit_span` is a no-op by default, but let's see what benchmarks say.

Note: I suggest calling the visitor `RemoveSpans` (plural), not `RemoveSpan`, because this will remove *all* spans, not just one.

**Comment by overlookmotel** [comment]
Looks like #7816 has no perf impact. If Boshen is happy with it, it will solve the problem in this PR.

**Review by overlookmotel — COMMENTED** [review]
Could you please add a test where replacement contains statements e.g. `(() => { return 123; })()`?

File: crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs [hunk 1]
```diff
@@ -235,3 +236,57 @@ console.log(
...
+    let snapshot = visualizer.into_visualizer_text();
+    insta::assert_snapshot!("test_sourcemap", snapshot);
+}
+
+#[test]
+#[should_panic]
+fn test_complex() {
+    let config = ReplaceGlobalDefinesConfig::new(&[("__DEF__", "((() => {})())")]).unwrap();
+    test("__DEF__", "1", config.clone());
+}
```
Comment (hi-ogawa) [inline crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs:292]: It turned out using more complex expression seems to panic due to lack of scope information. This happens on main too. Here is a stacktrace


```js
---- plugins::replace_global_defines::test_complex stdout ----
thread 'plugins::replace_global_defines::test_complex' panicked at /home/hiroshi/code/others/oxc/crates/oxc_traverse/src/generated/walk.rs:2400:10:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: oxc_traverse::generated::walk::walk_arrow_function_expression
   5: oxc_traverse::generated::walk::walk_expression
   6: oxc_traverse::generated::walk::walk_parenthesized_expression
   7: oxc_traverse::generated::walk::walk_expression
   8: oxc_traverse::generated::walk::walk_call_expression
   9: oxc_traverse::generated::walk::walk_expression
  10: oxc_traverse::generated::walk::walk_parenthesized_expression
  11: oxc_traverse::generated::walk::walk_expression
  12: oxc_traverse::generated::walk::walk_expression_statement
  13: oxc_traverse::generated::walk::walk_statement
  14: oxc_traverse::generated::walk::walk_statements
  15: oxc_traverse::generated::walk::walk_program
  16: oxc_traverse::traverse_mut_with_ctx
  17: oxc_traverse::traverse_mut
  18: oxc_transformer::plugins::replace_global_defines::ReplaceGlobalDefines::build
  19: integrations::plugins::replace_global_defines::test
  20: integrations::plugins::replace_global_defines::test_complex
  21: integrations::plugins::replace_global_defines::test_complex::{{closure}}
  22: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Comment (overlookmotel) [reply crates/oxc_transformer/tests/integrations/plugins/replace_global_defines.rs]: Oh damn!

That's a separate bug, related to use of `CloneIn`, which removes scope, symbol and reference information from the AST snippet. We don't have a solution to that right now. https://github.com/oxc-project/oxc/issues/4804

I'm going to merge this now and remove the panicking test in a follow-on.

Sorry I wasted your time with that red herring!

**Pull request merged**